metals/src/main/scala/scala/meta/internal/builds/MillDigest.scala

```diff
@@ -0,0 +1,98 @@
+package scala.meta.internal.builds
+import scala.util.control.NonFatal
+import scala.meta.tokens.Token
+import java.nio.file.Paths
+import scala.meta.io.AbsolutePath
+import scala.collection.mutable
+import java.security.MessageDigest
+import scala.meta.internal.mtags.MtagsEnrichments._
+
+object MillDigest extends Digestable {
+  override protected def digestWorkspace(
+      workspace: AbsolutePath,
+      digest: MessageDigest
+  ): Boolean = {
+    def analyzeBuildScript(file: AbsolutePath): Boolean = {
+      val imported = findImportedScripts(file)
+      imported.forall(script => analyzeBuildScript(script)) && Digest
+        .digestFile(file, digest)
+    }
+    val buildFile = workspace.resolve("build.sc")
+    analyzeBuildScript(buildFile)
+  }
+
+  private case class ImportLinesAcc(
+      var hadImport: Boolean = false,
+      var hadFile: Boolean = false,
+      var insideBraces: Boolean = false,
+      var skip: Boolean = false,
+      var allPaths: mutable.ListBuffer[AbsolutePath] = new mutable.ListBuffer,
+      var current: String = "",
+      var prefix: String = ""
+  ) {
+    def processingFileImport = hadFile && hadImport
+  }
+
+  // find all `import $file.path`
+  private def findImportedScripts(
+      file: AbsolutePath
+  ): List[AbsolutePath] = {
+    try {
+      val input = file.toInput
+      val tokens = input.tokenize.get.tokens
+      val acc = ImportLinesAcc()
+      tokens.toList.foreach {
+        case token =>
+          token match {
```

Comment:
I'm not a huge fan of the implementation details in this mini-parser. I have a feeling it can be simplified a bit, but if it works as it is then I'm happy 😊

Reply:
I think I am fine with the implementation now, but if you have any suggestions let me know. Marek wrote something trying to prove that it can be simpler, but turns out it is of a similar complexity.